### ISSUE TITLE
Connect-Safeguard with 2FA errors with "Code was already redeemed"

### ISSUE THREAD

**Issue by gilesmarshall** [OneIdentity/safeguard-ps#197]
Our safeguard appliance is configured for 2FA (Okta) so I'm using the -GUI switch with Connect-Safeguard.

Sometimes I cannot connect and get the following error:
Invoke-RestMethod : {"Error":"invalid_request","error_description":"Code was already redeemed.","success":"false"}

I have not been able to narrow down what causes this to work or fail.  Just trying again at some point and it works (or fails again).
The only references to this type of error I've found when searching Google is for other (non-safeguard) modules.  And the suggested fix is an update to the module.

I've tried with and without the -Insecure switch. 

Any suggestions?

**Comment by petrsnd** [comment]
Hey @gilesmarshall ,

Sorry for the slow response. I'd like to set up your scenario so that I can reproduce the issue internally.

How do you have your user set up? Could you run the following command for your user and post the output?

```PowerShell
PS> Get-SafeguardUser <username> -Fields UserName,IdentityProviderName,IdentityProviderTypeReferenceName,PrimaryAuthenticationProviderName,PrimaryAuthenticationProviderTypeReferenceName,SecondaryAuthenticationProviderName,SecondaryAuthenticationProviderTypeReferenceName
```

**Comment by gilesmarshall** [comment]
Hi @petrsnd thanks for the reply.
User is configured as below (username changed for obvious reasons)

`UserName                                         : AnAlias`
`IdentityProviderName                             : Active Directory`
`IdentityProviderTypeReferenceName                : ActiveDirectory`
`PrimaryAuthenticationProviderName                : OKTA`
`PrimaryAuthenticationProviderTypeReferenceName   : ExternalFederation`
`SecondaryAuthenticationProviderName              : `
`SecondaryAuthenticationProviderTypeReferenceName : Unknown`

To add to the error message, this is the line where it errors:

`Modules\safeguard-ps\2.9.379\safeguard-ps.psm1:152 char:5`

We are still running on safeguard v2.7 but I updated the modules to see if it fixed anything :-)

**Comment by Kevin-Andrew** [comment]
@gilesmarshall , can you run the `Connect-Safeguard` command with the `-Verbose` option and give us back the results.  If you can't make it fail again, you can still show us a successful output as well.  For example:
```
PS C:\WINDOWS\system32> Connect-Safeguard -gui 10.5.33.107 -verbose
VERBOSE: Configuring SSL version support to be secure
VERBOSE: POST https://10.5.33.107/RSTS/oauth2/token with -1-byte payload
VERBOSE: received 1511-byte response of content type application/json; charset=utf-8
VERBOSE: Calling Safeguard LoginResponse service...
VERBOSE: POST https://10.5.33.107/service/core/v3/Token/LoginResponse with -1-byte payload
VERBOSE: received 1008-byte response of content type application/json; charset=utf-8
VERBOSE: Setting up the SafeguardSession variable
Login Successful.
PS C:\WINDOWS\system32>
```

**Comment by Kevin-Andrew** [comment]
@gilesmarshall , are you using Safeguard in a cluster of appliances?  If so, how many nodes do you have in the cluster?  Are you using a load balancer in front of the cluster?

**Comment by gilesmarshall** [comment]
Yes, we have a 5 node cluster.
There is a load balancer, but I am connecting directly to the IP address of the Primary device.

**Comment by gilesmarshall** [comment]
Output as follows
`Connect-Safeguard -Appliance $SGappliance -gui -Verbose`
`VERBOSE: Configuring SSL version support to be secure`
`VERBOSE: Adding the PSType for rSTS Web form interaction`
`VERBOSE: POST https://<SG_Appliance_FQDN>/RSTS/oauth2/token with -1-byte payload`
`Invoke-RestMethod : {"error":"invalid_request","error_description":"Code was already redeemed.","success":false}`
`At C:...\Modules\safeguard-ps\2.9.379\safeguard-ps.psm1:152 char:5`
`+     Invoke-RestMethod -Method POST -Headers @{`
`+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~`
`    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-RestMethod], WebException`
`    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeRestMethodCommand`

Also tried with Insecure switch with the same result.
`Connect-Safeguard -Appliance $SGappliance -gui -Verbose -Insecure`
`VERBOSE: Configuring SSL version support to be secure`
`VERBOSE: Disabling SSL on Windows platform`
`VERBOSE: Adding the PSType for SSL trust override`
`VERBOSE: Adding the trust everything callback`
`VERBOSE: POST https://<SG_Appliance_FQDN>/RSTS/oauth2/token with -1-byte payload`
`VERBOSE: Enabling SSL on Windows platform`
`VERBOSE: Removing the trust everything callback`
`Invoke-RestMethod : {"error":"invalid_request","error_description":"Code was already redeemed.","success":false}`
`etc`

**Comment by Kevin-Andrew** [comment]
@gilesmarshall , thank you for the additional information.  Based upon the output from the `-verbose` switch, I don't see the script appear to be making multiple calls to try and redeem the authorization code.  So hopefully we can rule that out.

The error you are getting should be logged on the server.  In addition, I believe it will have also logged your authentication.  Therefore, what we can do is the following:

1. Reproduce the error again.
2. Record the date and time of when you encountered the error.
3. Generate a Safeguard Support Bundle, from all nodes in the cluster.  No need to "Include Event Logs" when generating the Support Bundle.  (Hopefully you have permissions to do this.)
4. Open a Support Case with One Identity.  Include all the above information.  You can reference this issue and ask that the support representative contact me.

**Comment by gilesmarshall** [comment]
Thanks @Kevin-Andrew .
> 4. Open a Support Case with One Identity.  Include all the above information.  You can reference this issue and ask that the support representative contact me.

I have opened a new case with the requested information and referenced this GitHub issues thread

**Comment by Kevin-Andrew** [comment]
@gilesmarshall , we have received the Support Bundles that you sent us.  Thank you for that.

After reviewing the Support Bundles, we can see that, for example, in the error case you mentioned, the web requests are in fact going to different nodes of the cluster behind the firewall.

We believe there is a change we can make to the Safeguard-PS module to correct this problem.

We will get back to you with more information about an update as soon as possible.

**Comment by petrsnd** [comment]
@gilesmarshall 
You should be able to verify the fix for this issue in the safeguard-ps 2.10 release.  Stay tuned.  I'll try to get it out this week.

You can try out the pre-release earlier if you'd like to.
